### DISCUSSION TITLE
difficulties running cf-execd

### DISCUSSION THREAD

**Discussion by LucVereecken** [cfengine/core#5469] [General]
I'm having difficulties running cf-execd (CFEngine Core 3.21.0). The logfile it tries to create in /var/cfengine/outputs has too long a filename.
This is what shows up in the output:
error: Couldn't open '/var/cfengine/outputs/cf________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________node0420___________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________jukinet_fz_juelich_de__1710172596_Mon_Mar_11_16_56_36_2024_0x7ff5ba697740' - aborting exec. (fopen: File name too long)

there are obviously too many underscores in the part where it should just be the fully qualified node name. I couldn't find anything on google about this, and as far as I can tell this is something that occurs inside the cfengine code, so not something I can affect by any configuration setting.

**Comment by olehermanse** [comment]
Hello @LucVereecken, how did you install CFEngine? Did you build from source? Or installed one of our prebuilt packages? Enterprise or community?

**Reply by LucVereecken** [reply]
This is the community version. I have the same problem both with the prebuild package provided, as with the packaged rpm included on the opensuse Leap 15.5 repositories.  It occurs both on a recent (opensuse Leap 15.5) and a very old (suse 13.2) opensuse distribution. Later CFEngine versions also show this issues on those machines. I haven't tried to go back version by version to find one that doesn't have this yet, but ancient versions of CFEngine (e.g. 3.8) don't show this problem. I also haven't tried compiling from source.

**Reply by olehermanse** [reply]
@LucVereecken I see, what does `cf-agent` output as the fully qualified domain name?

```
cf-agent --show-evaluated-vars="default:sys.fqhost"
```

FYI: I've created [a bug report in our issue tracker](https://northerntech.atlassian.net/browse/CFE-4354).

**Reply by LucVereecken** [reply]
That's indeed the variable that has the many spaces in:

default:sys.flavour                      opensuse_leap_15                                             source=agent                  
default:sys.fqhost          [A lot of spaces here]                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          
             node0420.                    [ A lot of spaces here]                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      
             jukinet.fz-juelich.de inventory,source=agent,attribute_name=Host name
default:sys.host                         node0420                                                     inventory,source=agent,attribute_name=none
default:sys.uqhost                       node0420                                                     inventory,source=agent,attribute_name=none
default:control_common.domain            jukinet.fz-juelich.de                                        source=promise

the [a lot of spaces here] was added by me as copy-pasting this in the reply box didn't make it very clear. When I use the sys.fqhost elsewhere in my configuration (e.g. the executor mailsubject) that too will have the many spaces.

I included the sys.flavor to show this is happening on opensuse leap 15.5, as I would guess it has to do reading the machine configuration. I also included some related variables as well, but none of those show the excess spaces. There are classes called "node0420_jukinet_fz_juelich_de", "node0420", and "jukinet_fz_juelich.de", which suggests that not all of the fully qualified names internally have all these spaces, and that the subparts seem ok. There is also a statement during a verbose run saying " verbose: SET domain = jukinet.fz-juelich.de", i.e. also without excess spaces. All this is pointing towards an incorrect concatenation (well... or getting a fqhost with spaces but then splitting it on the period and removing the spaces, depending on the implementation).

Thanks for making a bug report in the issue tracker.

**Reply by LucVereecken** [reply]
More relevant info: the domain name is set in the promises.cf file, in body common control { domain => "jukinet.fz-juelich.de"; } so the domain name is probably not derived from e.g. the /etc/hostname file, even when that contains the fqdn. The jukinet.fz-juelich.de is not a "real", registered domain, i.e. it doesn't exist in the DNS records, only in static configuration within my internal network.

**Reply by olehermanse** [reply]
@LucVereecken Fixing the "surface issue" of the long filename should be easy, we can limit the `sys.fqhost` variable to 256 characters, and also strip the spaces. However, the underlying issue (why there's so many spaces in the domain name) might be harder, could you try to reproduce it on a fresh machine (cloud VM for instance) and send us some instructions for doing so?

**Reply by LucVereecken** [reply]
@olehermanse These are fresh machines, installed using autoyast and set up using cfengine (running cf-agent as part of the init scripts in the autoyast file). That's quite an elaborate setup to communicate.
In the machine, the  domain name is set in the /etc/hostname file, in the /etc/hosts file, and in the sysctl kernel.domainname parameter. 
For CFEngine, the domain name is set in the "body common control' in promises.cf (as already mentioned). Interestingly, when bootstrapping, I send an update.cf file that doesn't have the "domain" directive in the body common control, and for those runs a normal output file name is generated (but obviously without a domainname in the file name). This again points to a problem when concatenating the hostname and the domain, with the latter coming from the promises.cf file.
I'll see if i can narrow things down, e.g. a simpler system setup.

**Reply by LucVereecken** [reply]
@olehermanse I narrowed down in which version of CFEngine the problem starts:
cfengine-community-3.18.0-1.suse15.x86_64.rpm -> has it
cfengine-community-3.15.7-1.suse15.x86_64.rpm -> doesn't have it
I didn't check the 3.16 and 3.17 versions as there are no suse15 rpms available for these non-LTS distributions.

**Reply by olehermanse** [reply]
> I'll see if i can narrow things down, e.g. a simpler system setup.

@LucVereecken Were you able to set up a simpler system and write some instructions, so that we can reproduce the problem on a cloud VM?

**Comment by LucVereecken** [comment]
@olehermanse I narrowed down in which version of CFEngine the problem starts:
cfengine-community-3.18.0-1.suse15.x86_64.rpm -> has it
cfengine-community-3.15.7-1.suse15.x86_64.rpm  -> doesn't have it
I didn't check the 3.16 and 3.17 versions as there are no suse15 rpms available for these non-LTS distributions.

**Comment by LucVereecken** [comment]
@olehermanse I finally found time to create a test case, using Opensuse Leap 15.5. It' very simple:
Install Leap 15.5:
  - standard installation medium as provided by opensuse. I used the large installation medium, not the network based, but I suspect that is not critical.
  - Allow online repositories to be added (the proposal is good, i.e. all non-debug repositories). I did this as I
   assume CF-Engine isn't on the install disk
  - Role: Standard desktop with plasma
  - When configuration options are given, select software, search for “cfengine” and add packages “cfengine” and “cfengine-masterfiles3”. Alternatively, you can just use zypper to install it after the installation. Confirmation will also show a list of dependencies (all selected packages), just accept them.
  - I used timezone Europe -> Germany. though I doubt that matters.
  - The remaining few questions (user account,...) as as you want.

Once logged in as root, I first did a "cf-agent --bootstrap localhost", but that is not necessary. I then created skeleton files in /var/cfengine/inputs : 

File ; /var/cfengine/inputs/promises.cf
```
body common control
{

version => "0.1";
domain => "jukinet.fz-juelich.de";

bundlesequence => {
       "update"
       };

inputs => {
       "update.cf"
       };
```

File /var/cfengine/inputs/update.cf : 
```
bundle agent update {

}
```

As you can see, this actually does nothing. Running cf-promises gives no errors.

Then running "cf-execd –no-fork –once" yields
```
node0418:/var/cfengine/inputs # cf-execd --no-fork --once
  error: Couldn't open '/var/cfengine/outputs/cf________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________node0418___________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________jukinet_fz_juelich_de__1713453174_Thu_Apr_18_17_12_54_2024_0x7f52692f6740' - aborting exec. (fopen: File name too long)
```

Then, comment out the "domain" directive, run the cf-execd command again. This will run without errors.

**Reply by nickanderson** [reply]
I ran into this myself yesterday and I knew it felt familiar. Didn't recall this discussion but @larsewi mentioned it today.

Anyway, I created a ticket for this here: https://northerntech.atlassian.net/browse/CFE-4389 but then I rejected it in favor of https://northerntech.atlassian.net/browse/CFE-4053 which was more specifically about the problem with setting `domain` in body common control.
